lib/rust/ensogl/core/src/display/object/event.rs

```diff
@@ -18,12 +18,31 @@ use crate::display::object::instance::WeakInstance;
 /// is cancelled, or that the propagation cannot be cancelled. See docs of this module to learn
 /// more.
 #[allow(missing_docs)]
-#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
 pub enum State {
+    Running(Phase),
+    RunningNonCancellable(Phase),
+    // Event has been cancelled, but the event handler is still running.
+    Cancelled(Phase),
+    // Event has been cancelled and [`InstanceDef::emit_event_impl`] function has returned.
+    Stopped(Phase),
+    Finished,
```

Comment:
1. What is finished? It can be many things IMO, lack of dock makes it hard to understand.
2. If I understand the meaning correctly, I think the name `Stopped` should be refactored to `FinishedCancelled` to make it more obvious what is the distinction between the two last.